modules/schema-util/src/main/scala/edu/gemini/schema/util/SchemaStitcher.scala

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package edu.gemini.schema.util
+
+import cats.data.{Ior, NonEmptySet}
+import cats.effect.Sync
+import cats.kernel.Order
+import cats.parse.Parser.*
+import cats.parse.Rfc5234.{alpha, digit, wsp}
+import cats.parse.{Parser, Parser0}
+import cats.syntax.all.*
+import cats.{Applicative, Eq, Monad}
+import edu.gemini.grackle.*
+import eu.timepit.refined.types.string.NonEmptyString
+
+import scala.io.Source
+import scala.tools.nsc.io.Path
+
+// SchemaStitcher can build a grackle Schema from schema files that contains import statements of the form:
+// #import <type list> from <schema file name>
+// <type list> is a coma separated list of the times to import, or * to import all of them
+trait SchemaStitcher[F[_]] {
+  def build: F[Result[Schema]]
+}
+
+object SchemaStitcher {
+
+  private case class SchemaStitcherImpl[F[_]: Sync](
+    root:     Path,
+    resolver: SourceResolver[F]
+  ) extends SchemaStitcher[F] {
+
+    // Process to build the schema:
+    // 1. Build a schema dependency tree starting with the root schema.
+    // 2. Recursively collapse the tree into a List, with dependencies first
+    // 3. Merge all elements referencing the same schema. Only the left-most one is left, and its type list is replaced
+    //    by the union of all the others
+    // 4. Build the final schema text joining the string representation of each schema. Only the used type are put in
+    //    the string.
+    // 5. Build the Schema instance
+    override def build: F[Result[Schema]] = dependenciesTree(List.empty, root)
+      .map(x => collapseToList(AllElements, x))
+      .map(x => merge(x, List.empty))
+      .map(_.map(_.asString).mkString("\n"))
+      .map(Schema.apply(_))
```

Comment:
Is the schema parsed here? Would it fail if there are schema errors?

Reply:
If you are importing all types there is no schema parsing. But if only some types are imported, then the schema is parsed. At this stage it is possible to have undefined type errors for types defined in other schema. I was lazy, and simply ignored errors. But at the end errors will appear when processing the resulting schema string.